### PR TITLE
Modernize CMake, add tests, streamline entire build

### LINKED ISSUE #21: GitHub actions
Transition the CI services to use GitHub's facilities instead of AppVeyor/Travis

### PR DESCRIPTION
**Pull request by scivision**
fixes #21 

Use modern CMake approach:

* per target properties including C++17, include and libraries
* add tests for executables suitable for that
* make command line option to autobuild tests and examples rather than having to dive into directories
* add missing Winsock libraries for Windows
* allow building without OpenSSL

* add GitHub Actions CI for Linux, Mac, MinGW -- note, you may need to enable GitHub Actions under the "actions" tab of this GitHub repository

* add Meson build system

### CODE REVIEW

**Comment by Ybalrid** [comment]
Oh, this is really interesting

File: tests/socket/main.cpp [hunk 1]
```diff
@@ -8,7 +8,7 @@
 using namespace std::chrono_literals;
 namespace kn = kissnet;
 
-int main()
+int main(void)
```
Comment (Ybalrid) [inline tests/socket/main.cpp:11]: This is C++ not C, `(void)` doesn't actually make sense, as empty param already means "no parameters". We don't need command-line arguments, but I'm not against putting `int argc, char* argv[]` here instead, or leave as-is

File: tests/simple/main.cpp [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #include <iostream>
 #include <kissnet.hpp>
 
-int main()
+int main(void)
```
Comment (Ybalrid) [inline tests/simple/main.cpp:4]: Same comment as the other `int main()` function, see below

File: tests/error_handler/main.cpp [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #include <iostream>
 #include <kissnet.hpp>
 
-int main()
+int main(void)
```
Comment (Ybalrid) [inline tests/error_handler/main.cpp:4]: Same comment as the other `int main()`, see below

File: tests/endpoint/main.cpp [hunk 4]
```diff
@@ -1,7 +1,7 @@
 #include <iostream>
 #include <kissnet.hpp>
 
-int main()
+int main(void)
```
Comment (Ybalrid) [inline tests/endpoint/main.cpp:4]: same comment as the other `int main()`, see below

File: tests/acceptor/main.cpp [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 namespace kn = kissnet;
 
-int main()
+int main(void)
```
Comment (Ybalrid) [inline tests/acceptor/main.cpp:6]: same comment as the other `int main()`, see below

File: kissnet.cpp [hunk 6]
```diff
@@ -2,6 +2,6 @@
 #define KISSNET_USE_OPENSSL
 #include "kissnet.hpp"
 
-int main()
+int main(void)
```
Comment (Ybalrid) [inline kissnet.cpp:5]: same comment as the other `int main()`, see below

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,13 +1,34 @@
...
+CMAKE_MINIMUM_REQUIRED(VERSION 3.13)
+set(CMAKE_CONFIGURATION_TYPES "Release;RelWithDebInfo;Debug" CACHE STRING "Build type selections" FORCE)
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "default build type")
+endif()
+project(kissnet_ci
+  LANGUAGES CXX
+  DESCRIPTION "header-only C++ network interface"
+  HOMEPAGE_URL https://github.com/Ybalrid/kissnet
+  VERSION 0.2.0)
```
Comment (Ybalrid) [inline CMakeLists.txt:10]: I don't think we should put it out as version 0.2 just yet 😉 

Keep that at version 0.1, we bump to 0.2 when we will break user-facing API

File: tests/socket/main.cpp (hunk 1 above)
Comment (scivision) [reply tests/socket/main.cpp]: I got into this habit from users that set:

```sh
CFLAGS: "-Werror=unused-parameter -Werror=return-type -Werror=strict-prototypes"
CPPFLAGS: "-Werror=unused-parameter -Werror=return-type"
```

but yes it's unnecessary for C++.  Thanks!

**Comment by Ybalrid** [comment]
Thanks for this, I'm gonna merge it now

**Comment by Ybalrid** [comment]
I will probably add a visual studio build to the `ci` workflow a bit later, just for the sake of completeness

**Pull request merged**